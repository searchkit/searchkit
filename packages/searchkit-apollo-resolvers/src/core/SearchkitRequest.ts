import dataloader from 'dataloader'
import { Client, ClientOptions } from '@elastic/elasticsearch'
import HttpAgent, { HttpsAgent } from 'agentkeepalive'
import { SearchkitConfig } from '../resolvers'
import QueryManager from './QueryManager'
import { filterTransform } from './FacetsFns'

export interface SearchResponse<T> {
  took: number
  timed_out: boolean
  hits: {
    total: {
      value: number
    }
    max_score: number
    hits: Array<{
      _id: number
      _source: any
      fields?: any
      highlight?: any
      inner_hits?: any
      matched_queries?: string[]
      sort?: string[]
    }>
  }
  aggregations?: any
}

export const mergeESQueries = (queries) =>
  queries.reduce(
    (combinedQuery, partial) => ({
      ...combinedQuery,
      ...partial,
      aggs: {
        ...combinedQuery.aggs,
        ...partial.aggs
      }
    }),
    {}
  )

const keepaliveHttpsAgent = new HttpsAgent()
const keepaliveAgent = new HttpAgent()

export default class SearchkitRequest {
  private dataloader: any
  private client: Client

  constructor(private queryManager: QueryManager, private config: SearchkitConfig) {
    this.client = new Client({
      node: this.config.host,
      agent: () =>
        new URL(this.config.host).protocol === 'http:' ? keepaliveAgent : keepaliveHttpsAgent
    })

    this.dataloader = new dataloader(async (partialQueries) => {
      const baseQuery = {
        size: 0,
        ...(this.queryManager.hasQuery() && this.config.query
          ? { query: this.config.query.getFilter(this.queryManager) }
          : {}),
        post_filter: filterTransform(this.queryManager, this.config.facets)
      }

      const ESQuery = mergeESQueries([baseQuery, ...(partialQueries as any[])])
      const response = await this.executeQuery(ESQuery)

      return partialQueries.map(() => response)
    })
  }

  public async search(partialQuery): Promise<SearchResponse<any>> {
    return this.dataloader.load(partialQuery)
  }

  private async executeQuery(esQuery): Promise<SearchResponse<any>> {
    try {
      const response = await this.client.search<SearchResponse<any>>({
        index: this.config.index,
        body: esQuery
      })

      return response.body
    } catch (e) {
      console.log(e)
    }
  }
}
