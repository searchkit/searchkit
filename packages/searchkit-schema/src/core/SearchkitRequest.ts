import dataloader from 'dataloader'
import { Client } from '@elastic/elasticsearch'
import HttpAgent, { HttpsAgent } from 'agentkeepalive'
import { SearchkitConfig } from '../resolvers'
import ESQueryError from '../utils/ESQueryError'
import { BaseFacet } from '../facets'
import QueryManager from './QueryManager'
import { filterTransform } from './FacetsFns'

export interface SearchResponse<T> {
  took: number
  timed_out: boolean
  hits: {
    total: {
      value: number
    }
    max_score: number
    hits: Array<{
      _id: number
      _source: any
      fields?: any
      highlight?: any
      inner_hits?: any
      matched_queries?: string[]
      sort?: string[]
    }>
  }
  aggregations?: any
}

export type Query = {
  bool?: {
    must?: Array<Record<string, unknown>>
    should?: Array<Record<string, unknown>>
    filter?: Array<Record<string, unknown>>
  }
}

type BaseQuery = {
  size: number
  query?: Query
  post_filter?: Query
}

export const mergeESQueries = (queries) =>
  queries.reduce(
    (combinedQuery, partial) => ({
      ...combinedQuery,
      ...partial,
      aggs: {
        ...combinedQuery.aggs,
        ...partial.aggs
      }
    }),
    {}
  )

const keepaliveHttpsAgent = new HttpsAgent()
const keepaliveAgent = new HttpAgent()

export default class SearchkitRequest {
  private dataloader: any
  private client: Client

  constructor(
    private queryManager: QueryManager,
    private config: SearchkitConfig,
    private baseFilters: Array<Record<string, unknown>>,
    private facets: Array<BaseFacet>
  ) {
    this.client = new Client({
      node: this.config.host,
      agent: () =>
        new URL(this.config.host).protocol === 'http:' ? keepaliveAgent : keepaliveHttpsAgent
    })

    this.dataloader = new dataloader(async (partialQueries) => {
      const query = this.buildQuery(partialQueries)
      const esQuery = this.config.postProcessRequest ? this.config.postProcessRequest(query) : query
      const response = await this.executeQuery(esQuery)
      return partialQueries.map(() => response)
    })
  }

  public async search(partialQuery): Promise<SearchResponse<any>> {
    return this.dataloader.load(partialQuery)
  }

  private buildQuery(partialQueries): BaseQuery {
    const queryFilter =
      this.queryManager.hasQuery() && this.config.query
        ? this.config.query.getFilter(this.queryManager)
        : null

    const baseFiltersQuery = filterTransform(this.queryManager, this.config.filters)
    const combinedBaseFilters = [].concat(this.baseFilters, baseFiltersQuery?.bool?.must || [])
    const query: Query = queryFilter || (combinedBaseFilters.length > 0 ? {} : null)

    if (combinedBaseFilters.length) {
      if (query.bool) {
        Object.assign(query.bool, {
          filter: query.bool.filter?.length
            ? [].concat(query.bool.filter, combinedBaseFilters)
            : combinedBaseFilters
        })
      } else {
        Object.assign(query, { bool: { filter: combinedBaseFilters } })
      }
    }

    const postFilter = filterTransform(this.queryManager, this.facets)

    let highlight
    this.config.hits.highlightedFields?.forEach((field) => {
      if (!highlight) {
        highlight = { fields: {} }
      }
      if (typeof field == 'string') {
        highlight.fields[field] = {}
      } else {
        highlight.fields[field.field] = field.config
      }
    })

    const baseQuery = { size: 0 }

    return mergeESQueries(
      [
        baseQuery,
        query && { query },
        postFilter && { post_filter: postFilter },
        highlight && { highlight },
        ...(partialQueries as any[])
      ].filter(Boolean)
    )
  }

  private async executeQuery(esQuery): Promise<SearchResponse<any>> {
    const debugMode = process.env.DEBUG_MODE === "true"
    try {
      if (debugMode) console.log(JSON.stringify(esQuery, null, 2))
      const response = await this.client.search<SearchResponse<any>>({
        index: this.config.index,
        body: esQuery
      })

      return response.body
    } catch (e) {
      if (debugMode) console.log(JSON.stringify(e, null, 2))

      if (e.meta?.statusCode === 400) {
        throw new ESQueryError(
          `Elasticsearch query failed. Check your custom filters or configuration. Below is the ES Query`,
          esQuery
        )
      } else {
        throw e
      }
    }
  }
}
