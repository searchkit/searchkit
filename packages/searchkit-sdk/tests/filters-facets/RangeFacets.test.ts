import nock from 'nock'
import SearchkitRequest, {
  RangeFacet,
  DateRangeFacet,
  MultiMatchQuery,
  SearchkitConfig
} from '../../src'
import ResultsNoHitsMock from '../__mock-data__/Facets/results.json'

describe('Numeric Range + Date Range Facet Filters', () => {
  describe('Numeric Range', () => {
    const config: SearchkitConfig = {
      host: 'http://localhost:9200',
      index: 'movies',
      hits: {
        fields: ['actors', 'writers']
      },
      query: new MultiMatchQuery({ fields: ['actors', 'writers', 'title^4', 'plot'] }),
      facets: [
        new RangeFacet({
          field: 'imdbRating',
          label: 'IMDB Rating',
          identifier: 'imdbRating',
          range: {
            min: 0,
            max: 100,
            interval: 10
          }
        }),
        new DateRangeFacet({
          field: 'released',
          label: 'Released',
          identifier: 'released'
        })
      ]
    }

    it('Combination of min and max number range facets', async () => {
      const request = SearchkitRequest(config)
      request.setFilters([
        {
          identifier: 'imdbRating',
          min: 5,
          max: 10
        }
      ])

      let lastESRequest

      const scope = nock('http://localhost:9200')
        .post('/movies/_search')
        .reply(200, (uri, body) => {
          expect(body).toMatchSnapshot()
          lastESRequest = body
          return ResultsNoHitsMock
        })

      const response = await request.execute({
        facets: true
      })
      expect(response.facets.find(({ label }) => label === 'IMDB Rating')).toMatchSnapshot(
        'imdb rating values'
      )
      expect(response.summary.appliedFilters).toMatchInlineSnapshot(`
        Array [
          Object {
            "display": "RangeSliderFacet",
            "id": "imdbRating_5_10",
            "identifier": "imdbRating",
            "label": "IMDB Rating",
            "max": 10,
            "min": 5,
            "type": "NumericRangeSelectedFilter",
          },
        ]
      `)
      expect(lastESRequest.aggs).toMatchInlineSnapshot(`
        Object {
          "facet_bucket_all": Object {
            "aggs": Object {},
            "filter": Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "range": Object {
                      "imdbRating": Object {
                        "gte": 5,
                        "lte": 10,
                      },
                    },
                  },
                ],
              },
            },
          },
          "facet_bucket_imdbRating": Object {
            "aggs": Object {
              "imdbRating": Object {
                "histogram": Object {
                  "extended_bounds": Object {
                    "max": 100,
                    "min": 0,
                  },
                  "field": "imdbRating",
                  "interval": 10,
                  "min_doc_count": 0,
                },
              },
            },
            "filter": Object {
              "bool": Object {
                "must": Array [],
              },
            },
          },
          "facet_bucket_released": Object {
            "aggs": Object {},
            "filter": Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "range": Object {
                      "imdbRating": Object {
                        "gte": 5,
                        "lte": 10,
                      },
                    },
                  },
                ],
              },
            },
          },
        }
      `)
    })

    it('Combination of min and max date range facets', async () => {
      const request = SearchkitRequest(config)
      request.setFilters([
        {
          identifier: 'released',
          dateMin: '2012-12-18T00:00:00.000Z',
          dateMax: '2021-12-18T00:00:00.000Z'
        }
      ])

      let lastESRequest

      const scope = nock('http://localhost:9200')
        .post('/movies/_search')
        .reply(200, (uri, body) => {
          expect(body).toMatchSnapshot()
          lastESRequest = body
          return ResultsNoHitsMock
        })

      const response = await request.execute({
        facets: true
      })
      expect(response.facets.find(({ label }) => label === 'released')).toMatchSnapshot(
        'released values'
      )
      expect(response.summary.appliedFilters).toMatchInlineSnapshot(`
        Array [
          Object {
            "dateMax": "2021-12-18T00:00:00.000Z",
            "dateMin": "2012-12-18T00:00:00.000Z",
            "display": "DateRangeFacet",
            "id": "released_2012-12-18T00:00:00.000Z_2021-12-18T00:00:00.000Z",
            "identifier": "released",
            "label": "Released",
            "type": "DateRangeSelectedFilter",
          },
        ]
      `)
    })
  })
})
