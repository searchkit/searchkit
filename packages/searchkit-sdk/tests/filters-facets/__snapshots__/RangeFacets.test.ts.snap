// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Numeric Range + Date Range Facet Filters Numeric Range Combination of min and max date range facets 1`] = `
Object {
  "_source": Object {
    "includes": Array [
      "actors",
      "writers",
    ],
  },
  "aggs": Object {
    "facet_bucket_all": Object {
      "aggs": Object {},
      "filter": Object {
        "bool": Object {
          "must": Array [
            Object {
              "range": Object {
                "released": Object {
                  "gte": "2012-12-18T00:00:00.000Z",
                  "lte": "2021-12-18T00:00:00.000Z",
                },
              },
            },
          ],
        },
      },
    },
    "facet_bucket_imdbRating": Object {
      "aggs": Object {
        "imdbRating": Object {
          "histogram": Object {
            "extended_bounds": Object {
              "max": 100,
              "min": 0,
            },
            "field": "imdbRating",
            "interval": 10,
            "min_doc_count": 0,
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "must": Array [
            Object {
              "range": Object {
                "released": Object {
                  "gte": "2012-12-18T00:00:00.000Z",
                  "lte": "2021-12-18T00:00:00.000Z",
                },
              },
            },
          ],
        },
      },
    },
    "facet_bucket_released": Object {
      "aggs": Object {},
      "filter": Object {
        "bool": Object {
          "must": Array [],
        },
      },
    },
  },
  "from": 0,
  "post_filter": Object {
    "bool": Object {
      "must": Array [
        Object {
          "range": Object {
            "released": Object {
              "gte": "2012-12-18T00:00:00.000Z",
              "lte": "2021-12-18T00:00:00.000Z",
            },
          },
        },
      ],
    },
  },
  "size": 0,
  "sort": Array [
    Object {
      "_score": "desc",
    },
  ],
}
`;

exports[`Numeric Range + Date Range Facet Filters Numeric Range Combination of min and max date range facets: released values 1`] = `undefined`;

exports[`Numeric Range + Date Range Facet Filters Numeric Range Combination of min and max number range facets 1`] = `
Object {
  "_source": Object {
    "includes": Array [
      "actors",
      "writers",
    ],
  },
  "aggs": Object {
    "facet_bucket_all": Object {
      "aggs": Object {},
      "filter": Object {
        "bool": Object {
          "must": Array [
            Object {
              "range": Object {
                "imdbRating": Object {
                  "gte": 5,
                  "lte": 10,
                },
              },
            },
          ],
        },
      },
    },
    "facet_bucket_imdbRating": Object {
      "aggs": Object {
        "imdbRating": Object {
          "histogram": Object {
            "extended_bounds": Object {
              "max": 100,
              "min": 0,
            },
            "field": "imdbRating",
            "interval": 10,
            "min_doc_count": 0,
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "must": Array [],
        },
      },
    },
    "facet_bucket_released": Object {
      "aggs": Object {},
      "filter": Object {
        "bool": Object {
          "must": Array [
            Object {
              "range": Object {
                "imdbRating": Object {
                  "gte": 5,
                  "lte": 10,
                },
              },
            },
          ],
        },
      },
    },
  },
  "from": 0,
  "post_filter": Object {
    "bool": Object {
      "must": Array [
        Object {
          "range": Object {
            "imdbRating": Object {
              "gte": 5,
              "lte": 10,
            },
          },
        },
      ],
    },
  },
  "size": 0,
  "sort": Array [
    Object {
      "_score": "desc",
    },
  ],
}
`;

exports[`Numeric Range + Date Range Facet Filters Numeric Range Combination of min and max number range facets: imdb rating values 1`] = `
Object {
  "display": "RangeSliderFacet",
  "entries": Array [
    Object {
      "count": 3120,
      "label": "0",
    },
    Object {
      "count": 521,
      "label": "10",
    },
    Object {
      "count": 473,
      "label": "20",
    },
    Object {
      "count": 58,
      "label": "30",
    },
    Object {
      "count": 28,
      "label": "40",
    },
    Object {
      "count": 98,
      "label": "50",
    },
    Object {
      "count": 28,
      "label": "60",
    },
    Object {
      "count": 48,
      "label": "70",
    },
    Object {
      "count": 68,
      "label": "80",
    },
    Object {
      "count": 38,
      "label": "90",
    },
    Object {
      "count": 28,
      "label": "100",
    },
  ],
  "identifier": "imdbRating",
  "label": "IMDB Rating",
  "type": "RangeFacet",
}
`;
